packages/common/client-utils/src/layerCompat.ts

```diff
@@ -0,0 +1,144 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Result of a layer compatibility check - whether a layer is compatible with another layer.
+ * @internal
+ */
+export type LayerCompatCheckResult =
+	| { readonly isCompatible: true }
+	| {
+			readonly isCompatible: false;
+			/**
+			 * Whether the generation of the layer is compatible with the other layer.
+			 */
+			readonly isGenerationCompatible: boolean;
+			/**
+			 * The features that are required by the layer but are not supported by the other layer. This will only
+			 * be set if there are unsupported features.
+			 */
+			readonly unsupportedFeatures: string[] | undefined;
+	  };
+
+/**
+ * @internal
+ */
+export const ICompatibilityDetails: keyof IProvideCompatibilityDetails =
+	"ICompatibilityDetails";
+
+/**
+ * @internal
+ */
+export interface IProvideCompatibilityDetails {
+	readonly ICompatibilityDetails: ICompatibilityDetails;
+}
+
+/**
+ * This interface is used to communicate the compatibility details of a layer to another layer.
+ * @internal
+ */
+export interface ICompatibilityDetails extends IProvideCompatibilityDetails {
+	/**
+	 * A list of features supported by the layer at a particular layer boundary. This is used to check if these
+	 * set of features satisfy the requirements of another layer.
+	 */
+	supportedFeatures: Set<string>;
+	/**
+	 * The generation of the layer. The other layer at the layer boundary uses this to check if this satisfies
+	 * the minimum generation it requires to be compatible.
+	 */
+	generation: number;
+	/**
+	 * The package version of the layer. When an incompatibility is detected, this is used to provide more context
+	 * on what the versions of the incompatible layers are.
+	 */
+	pkgVersion: string;
+}
+
+/**
+ * Manages the compatibility details of a layer.
+ * @internal
+ */
+export class LayerCompatibilityManager implements ICompatibilityDetails {
+	public get ICompatibilityDetails(): ICompatibilityDetails {
```

Comment:
i feel like this should take an ICompatailityDetails, rather than be ICompatailityDetails. i would keep the data seperate from the implementation. the below them can also just become a free function, rather than a class

Reply:
I created this class to make implementing `ICompatibilityDetails` which extended `IProvideCompatibilityDetails` easier since it would be done in multiple places. However, changing it to  `ICompatibilityDetails extends Partial<IProvideCompatibilityDetails>` as per your suggestion made things much easier. I removed this class now.